sql/catalog/lease: TestLeaseManagerLockedTimestampBasic failed

sql/catalog/lease.TestLeaseManagerLockedTimestampBasic [failed](https://mesolite.cluster.engflow.com/invocations/default/28c0fc1e-f827-47e9-aabd-223e579ea779?testReportRun=3&testReportShard=4&testReportAttempt=1#targets-Ly9wa2cvc3FsL2NhdGFsb2cvbGVhc2U6bGVhc2VfdGVzdA==) on master @ [0cb591927ad6b9b2d547d71552716c5515618cec](https://github.com/cockroachdb/cockroach/commits/0cb591927ad6b9b2d547d71552716c5515618cec):


```
=== RUN   TestLeaseManagerLockedTimestampBasic
    test_log_scope.go:172: test logs captured to: outputs.zip/logTestLeaseManagerLockedTimestampBasic3956162137
    test_log_scope.go:82: use -show-logs to present logs inline
    test_server_shim.go:635: DRPC is enabled (via testserver factory defaults)
    lease_internal_test.go:1860: 
        	Error Trace:	pkg/sql/catalog/lease/lease_internal_test.go:1860
        	Error:      	Not equal: 
        	            	expected: 0x2
        	            	actual  : 0x1
        	Test:       	TestLeaseManagerLockedTimestampBasic
        	Messages:   	new version should be visible now
    panic.go:615: -- test log scope end --
test logs left over in: outputs.zip/logTestLeaseManagerLockedTimestampBasic3956162137
--- FAIL: TestLeaseManagerLockedTimestampBasic (0.32s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>run=3</code>
 - <code>shard=4</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/sql-foundations
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestLeaseManagerLockedTimestampBasic.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-62467

sql/catalog/lease.TestLeaseManagerLockedTimestampBasic [failed](https://mesolite.cluster.engflow.com/invocations/default/b0b57007-f398-4245-824d-5b9028f6afaa?testReportRun=7&testReportShard=4&testReportAttempt=1#targets-Ly9wa2cvc3FsL2NhdGFsb2cvbGVhc2U6bGVhc2VfdGVzdA==) on master @ [7e489320238b6feb3fcf94afa5727b7dff7cf51f](https://github.com/cockroachdb/cockroach/commits/7e489320238b6feb3fcf94afa5727b7dff7cf51f):


```
=== RUN   TestLeaseManagerLockedTimestampBasic
    test_log_scope.go:172: test logs captured to: outputs.zip/logTestLeaseManagerLockedTimestampBasic1623019974
    test_log_scope.go:82: use -show-logs to present logs inline
    test_server_shim.go:635: DRPC is enabled (via testserver factory defaults)
    lease_internal_test.go:1860: 
        	Error Trace:	pkg/sql/catalog/lease/lease_internal_test.go:1860
        	Error:      	Not equal: 
        	            	expected: 0x2
        	            	actual  : 0x1
        	Test:       	TestLeaseManagerLockedTimestampBasic
        	Messages:   	new version should be visible now
    panic.go:615: -- test log scope end --
test logs left over in: outputs.zip/logTestLeaseManagerLockedTimestampBasic1623019974
--- FAIL: TestLeaseManagerLockedTimestampBasic (0.33s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>run=7</code>
 - <code>shard=4</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestLeaseManagerLockedTimestampBasic.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

sql/catalog/lease.TestLeaseManagerLockedTimestampBasic [failed](https://mesolite.cluster.engflow.com/invocations/default/181a8809-3f58-4b1d-a018-802cc147ff5e?testReportRun=1&testReportShard=4&testReportAttempt=1#targets-Ly9wa2cvc3FsL2NhdGFsb2cvbGVhc2U6bGVhc2VfdGVzdA==) on master @ [f9bcbc4fdf0ba5236f6312c00becf303c29f98f9](https://github.com/cockroachdb/cockroach/commits/f9bcbc4fdf0ba5236f6312c00becf303c29f98f9):


```
=== RUN   TestLeaseManagerLockedTimestampBasic
    test_log_scope.go:172: test logs captured to: outputs.zip/logTestLeaseManagerLockedTimestampBasic4248302319
    test_log_scope.go:82: use -show-logs to present logs inline
    test_server_shim.go:635: DRPC is enabled (via testserver factory defaults)
    lease_internal_test.go:1860: 
        	Error Trace:	pkg/sql/catalog/lease/lease_internal_test.go:1860
        	Error:      	Not equal: 
        	            	expected: 0x2
        	            	actual  : 0x1
        	Test:       	TestLeaseManagerLockedTimestampBasic
        	Messages:   	new version should be visible now
    panic.go:615: -- test log scope end --
test logs left over in: outputs.zip/logTestLeaseManagerLockedTimestampBasic4248302319
--- FAIL: TestLeaseManagerLockedTimestampBasic (0.34s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>run=1</code>
 - <code>shard=4</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestLeaseManagerLockedTimestampBasic.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

--- a/pkg/sql/catalog/lease/lease_internal_test.go
+++ b/pkg/sql/catalog/lease/lease_internal_test.go
@@ -1753,6 +1753,7 @@ func TestLeaseManagerLockedTimestampBasic(t *testing.T) {
 	// up.
 	WaitForInitialVersion.Override(ctx, &st.SV, false)
 	var tableID atomic.Int64
+	var expectedVersion atomic.Int64
 	srv, db, _ := serverutils.StartServer(
 		t, base.TestServerArgs{
 			// Avoid using tenants since async tenant migration steps can acquire
@@ -1766,8 +1767,13 @@ func TestLeaseManagerLockedTimestampBasic(t *testing.T) {
 						if !blockUpdates.Load() || descriptor == nil {
 							return nil
 						}
-						id, _, _, _, _ := descpb.GetDescriptorMetadata(descriptor)
+						id, version, _, _, _ := descpb.GetDescriptorMetadata(descriptor)
 						if id != descpb.ID(tableID.Load()) {
+							return nil
+						}
+						// Skip events for versions we've already seen. The rangefeed
+						// may deliver older versions after blockUpdates is set.
+						if int64(version) < expectedVersion.Load() {
 							return nil
 						}
 						ts, err := descpb.GetDescriptorModificationTime(descriptor)
@@ -1797,13 +1803,6 @@ func TestLeaseManagerLockedTimestampBasic(t *testing.T) {
 	var id int
 	r.QueryRow(t, "SELECT 't1'::REGCLASS::OID;").Scan(&id)
 	tableID.Store(int64(id))
-	blockUpdates.Store(true)
-
-	grp := ctxgroup.WithContext(context.Background())
-	grp.GoCtx(func(ctx context.Context) error {
-		_, err := db.Exec("ALTER TABLE t1 ADD COLUMN n2 int")
-		return err
-	})
 
 	lm := s.LeaseManager().(*Manager)
 	getDescriptorVersion := func(ts ReadTimestamp) descpb.DescriptorVersion {
@@ -1846,6 +1845,17 @@ func TestLeaseManagerLockedTimestampBasic(t *testing.T) {
 	}
 	defer releaseTS()
 	initialVersion := getDescriptorVersion(ts)
+	// Set expectedVersion so the hook only fires for the ALTER TABLE's new
+	// version, not for stale rangefeed deliveries of the current version.
+	expectedVersion.Store(int64(initialVersion + 1))
+	blockUpdates.Store(true)
+
+	grp := ctxgroup.WithContext(context.Background())
+	grp.GoCtx(func(ctx context.Context) error {
+		_, err := db.Exec("ALTER TABLE t1 ADD COLUMN n2 int")
+		return err
+	})
+
 	// The old version will still be cached as long as this timestamp is in use.
 	// Even if we released the leases already.
 	afterTS := <-updateCh

